Azure/azure-cli
# {App Service} `az webapp up`: include project references for dotnet w…

Include project references for dotnet webapp (#16399)

**Related command**
`az webapp up`

**Description**

\[Why this PR?\]
While learning some features of `azure cli` on MSLearn, I tried to execute `az webapp up` for my dotnet6 project and an error has occured. I found the relevent issue #16399.

I well aware that this PR could not cover all cases, but it may be the first step, by allowing to deploy a wide range of projects.
I'm a dotnet developer, and it's common to see the webapp projects using project references from the same solution.

I was curious about azure-cli, so I forked the project and I would like to suggest an adaptation with this PR.
It was just fun to analyze and make it. I hope it will help :)

\[What is changed?\]
I'm not a Python developer, I tried to make it simple with slight changes to the existing code. 

So, in case of dotnet webapp only, this feature allows to append in the zip (created  in `zip_contents_from_dir`) the project references found in .csproj files, including transitive project references.

There are 5 steps, executed only if there is at least one project reference in the webapp project:
1. create a temporary .zip file (.zip.tmp);
1. append the project references of the webapp to the .zip.tmp file and replace the content of the webapp .csproj;
1. append the transitive dependencies to the .zip.tmp file;
1. copy the content from the original zip to the .zip.tmp file;
1. replace the original .zip file with the .zip.tmp file.

The project references are included into a directory, with the name defined in a constant `DOTNET_REFERENCES_DIR_IN_ZIP` (I initialized it as `.az-references`).

To make the `dotnet run` command successful, it is necessary to update the .csproj by changing the path of the original project references.
The .csproj contains a project references to `../ClassLibrary{N}`, so it needs to be changed with `.az-references/ClassLibrary{N}`.

In order to make it simple, I don't update the .csproj of the project references themselves.

Finally, in case of exception (e.g: permissions), the behaviour still will be to use the original .zip file.

\[What is the effect?\]

The .zip archive contains what is needed to run the project.

\[Personal note\]

I'm aware that this PR is only intended to deal with project references at the same level in a directory.

Example: the function `FunctionApp1` and the webapp `WebApplication1` share the same dependencies, and all are is the same directory `my-webapp`. 
```
my-webapp
|   my-webapp.sln
|
+---ClassLibrary1
|   |   ClassLibrary1.csproj
|
+---ClassLibrary2
|   |   ClassLibrary2.csproj
|
+---ClassLibrary3
|   |   ClassLibrary3.csproj
|
+---FunctionApp1
|   |   FunctionApp1.csproj
|
\---WebApplication1
    |   WebApplication1.csproj
```

Also, I'm not really satisfied with the "zip replacement", but it is the only way I found to replace the existing .csproj into the zip archive.

**Testing Guide**

I followed the developper guide, so I ran `azdev [style/linter/test] appservice`

I added one test with a sample dotnet project using project references, but it just assert that the zip archive is made.

I don't think that it is the responsability of azure-cli to ensure that the webapp run successfully, because it depends of `/api/zipDeploy`, and I don't know If I can test it in real condition. 
So, I asserted that the test case in the issue is working, and also the sample used in the test.
